symbolic/src/Data/Macaw/Symbolic/Stack.hs

```diff
@@ -103,11 +110,127 @@ createArrayStack bak mem slots sz = do
   (base, mem1) <- mallocStack bak mem sz
   mem2 <- CLM.doArrayStore bak mem1 base stackAlign arr sz
 
-  end <- CLM.doPtrAddOffset bak mem2 base sz
+  off <- WI.bvSub sym sz =<< WI.bvOne sym ?ptrWidth
+  end <- CLM.doPtrAddOffset bak mem2 base off
   let ptrBytes = WI.intValue ?ptrWidth `div` 8
   let slots' = fromIntegral (getExtraStackSlots slots)
   let slotsBytes = slots' * ptrBytes
   slotsBytesBv <- WI.bvLit sym ?ptrWidth (BVS.mkBV ?ptrWidth slotsBytes)
   top <- CLM.ptrSub sym ?ptrWidth end slotsBytesBv
 
   pure (ArrayStack base top arr, mem2)
+
+-- | Allocate stack space by subtracting from the stack pointer
+--
+-- This is only suitable for use with architectures/ABIs where the stack grows
+-- down.
+allocStackSpace ::
+  C.IsSymInterface sym =>
+  CLM.HasPtrWidth w =>
+  sym ->
+  CLM.LLVMPtr sym w ->
+  WI.SymBV sym w ->
+  IO (CLM.LLVMPtr sym w)
+allocStackSpace sym = CLM.ptrSub sym ?ptrWidth
+
+-- | Align the stack pointer to a particular 'CLD.Alignment'.
+--
+-- This is only suitable for use with architectures/ABIs where the stack grows
+-- down.
+alignStackPointer ::
+  C.IsSymInterface sym =>
+  CLM.HasPtrWidth w =>
+  sym ->
+  CLD.Alignment ->
+  CLM.LLVMPtr sym w ->
+  IO (CLM.LLVMPtr sym w)
+alignStackPointer sym align orig@(CLM.LLVMPointer blk off) = do
+  let alignInt = 2 ^ naturalToInteger (CLD.alignmentToExponent align)
+  if alignInt == 1
+  then pure orig
+  else do
+    -- Because the stack grows down, we can align it by simply ANDing the stack
+    -- pointer with a mask with some amount of 1s followed by some amount of 0s.
+    let mask = BVS.complement ?ptrWidth (BVS.mkBV ?ptrWidth (alignInt - 1))
+    maskBv <- WI.bvLit sym ?ptrWidth mask
```

Comment:
I think this assumes that the required alignment is a power of two. If so, it would be worth documenting this.

Reply:
Isn't alignment always a power of 2? Note the definition of `alignInt`:
```haskell
  let alignInt = 2 ^ naturalToInteger (CLD.alignmentToExponent align)
```
The Haddocks on `Alignment` state:

> An Alignment represents a number of bytes that must be a power of two.

Reply:
Oops, quite right. Please ignore me.